demos/common/processing/async_pipeline.h

```diff
@@ -0,0 +1,150 @@
+/*
+// Copyright (C) 2018-2020 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+*/
+
+#define MOVING_AVERAGE_SAMPLES 5
+
+#pragma once
+#include <string>
+#include <deque>
+#include <map>
+#include <condition_variable>
+#include "config_factory.h"
+#include "requests_pool.h"
+#include "results.h"
+#include "model_base.h"
+
+/// This is base class for asynchronous pipeline
+/// Derived classes should add functions for data submission and output processing
+class PipelineBase
+{
+public:
+    struct PerformanceInfo
+    {
+        int64_t framesCount = 0;
+        std::chrono::steady_clock::duration latencySum = std::chrono::steady_clock::duration::zero();
+        std::chrono::steady_clock::duration lastInferenceLatency = std::chrono::steady_clock::duration::zero();
+        std::chrono::steady_clock::time_point startTime;
+        double movingAverageLatencyMs;
+        uint32_t numRequestsInUse;
+        double movingAverageFPS;
+        double FPS=0;
+
+        double getTotalAverageLatencyMs() const {
+            return ((double)std::chrono::duration_cast<std::chrono::milliseconds>(latencySum).count()) / framesCount;
+        }
+
+        double getLastInferenceLatencyMs() const {
+            return ((double)std::chrono::duration_cast<std::chrono::milliseconds>(lastInferenceLatency).count());
+        }
+
+    };
+
+    struct PerformanceInternalCounters
+    {
+        long long latenciesMs[MOVING_AVERAGE_SAMPLES] = {};
+        std::chrono::steady_clock::time_point retrievalTimestamps[MOVING_AVERAGE_SAMPLES] = {};
+        long long movingLatenciesSumMs = 0;
+        int currentIndex = 0;
+    };
+
+public:
+    /// Loads model and performs required initialization
+    /// @param modelInstance pointer to model object. Object it points to should not be destroyed manually after passing pointer to this function.
```

Comment:
This sounds like it should be passed as a `unique_ptr`.

Reply:
changed